x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/lookup/CachedDirectoryReader.java

```diff
@@ -0,0 +1,138 @@
+        return new CachedDirectoryReader(in);
+    }
+
+    @Override
+    public CacheHelper getReaderCacheHelper() {
+        return in.getReaderCacheHelper();
+    }
+
+    static class CachedLeafReader extends FilterLeafReader {
+        final Map<String, NumericDocValues> docValues = new HashMap<>();
+
+        CachedLeafReader(LeafReader in) {
+            super(in);
+        }
+
+        @Override
+        public NumericDocValues getNumericDocValues(String field) throws IOException {
+            NumericDocValues dv = super.getNumericDocValues(field);
+            if (dv == null) {
+                return null;
+            }
+            return new CachedNumericDocValues(docId -> docValues.compute(field, (k, curr) -> {
+                if (curr == null || curr.docID() > docId) {
+                    return dv;
+                }
+                return curr;
+            }));
+        }
+
+        @Override
+        public Terms terms(String field) throws IOException {
+            Terms terms = super.terms(field);
+            if (terms == null) {
+                return null;
+            }
+            return new FilterTerms(terms) {
+                @Override
+                public TermsEnum iterator() throws IOException {
+                    return in.iterator();
+                }
+            };
+        }
+
+        @Override
+        public CacheHelper getCoreCacheHelper() {
+            return in.getCoreCacheHelper();
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            return in.getCoreCacheHelper();
+        }
+    }
+
+    static class CachedNumericDocValues extends NumericDocValues {
+        private NumericDocValues delegate = null;
+        private final IntFunction<NumericDocValues> fromCache;
+
+        CachedNumericDocValues(IntFunction<NumericDocValues> fromCache) {
+            this.fromCache = fromCache;
+        }
+
+        NumericDocValues getDelegate(int docID) {
+            if (delegate == null) {
+                delegate = fromCache.apply(docID);
+            }
+            return delegate;
+        }
+
+        @Override
+        public long longValue() throws IOException {
+            return getDelegate(-1).longValue();
+        }
+
+        @Override
+        public boolean advanceExact(int target) throws IOException {
+            return getDelegate(target).advanceExact(target);
+        }
+
+        @Override
+        public int advance(int target) throws IOException {
+            return getDelegate(target).nextDoc();
+        }
+
+        @Override
+        public int docID() {
+            return getDelegate(-1).docID();
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return getDelegate(-1).nextDoc();
+        }
+
+        @Override
+        public long cost() {
+            return fromCache.apply(DocIdSetIterator.NO_MORE_DOCS).cost();
```

Comment:
Worth a comment here about how it's safe to discard the delegate in this case because the cost will always be the same.

Reply:
So whether or not we use the delegate after this call to `cost` it's fine - we're going to have returned the same cost.

Reply:
I don't think we can use the delegate for the cost estimate.